### PR TITLE
Add `then_with_cusolver` support for HIP

### LINKED ISSUE #48: Enable hipsolver functionality
We currently only have a macro translation layer for basic CUDA functionality and cuBLAS functionality to HIP equivalents. cuSOLVER is currently CUDA-only.

### PR DESCRIPTION
**Pull request by aurianer**
Maybe fixes #48

TODO:
- [x] Test pika on ault
- [x] Test with DLA-Future
- [ ] Add `rocblas` and `rocsolver` dependency to spack package (this will be done for 0.5.0)

### CODE REVIEW

File: cmake/pika_setup_hip.cmake [hunk 1]
```diff
@@ -21,6 +22,10 @@ if(PIKA_WITH_HIP AND NOT TARGET roc::hipblas)
   # Setup hipblas (creates roc::hipblas)
   find_package(hipblas REQUIRED HINTS $ENV{HIPBLAS_ROOT} CONFIG)
 
+  # Tmp! Until hipSOLVER is complete for DLA-Future
+  find_package(rocblas REQUIRED HINTS $ENV{ROCBLAS_ROOT})
+  find_package(rocsolver REQUIRED HINTS $ENV{ROCSOLVER_ROOT})
+
```
Comment (albestro) [inline cmake/pika_setup_hip.cmake:28]: Together with this, it is required also to add `roc::rocblas` and `roc::rocsolver` here

https://github.com/pika-org/pika/blob/a5c06c5146b3a1618796b50272dc8fdc063a893a/libs/pika/async_cuda/CMakeLists.txt#L50-L51

**Comment by msimberg** [comment]
I've updated this to put all the test-specific {cu,roc,hip}{blas,solver} algorithm translation macros in the tests themselves. We only require the basic handles, streams, and other core functionality for `libpika` itself. I'll go ahead with this as is, but further cleanup and updates may be needed in the future.

bors merge

**Comment by msimberg** [comment]
bors merge

**Comment by msimberg** [comment]
bors merge

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try-

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors merge

**Pull request merged**